File: DevEnvironment/CANMessageGenerator/src/main/java/canmessagegenerator/core/CANMessageId.java
package canmessagegenerator.core;

import canmessagegenerator.interfaces.ICANMessageId;

/**
 * The CANMessageId class implements an id of a controller area network message.
 */
public final class CANMessageId implements ICANMessageId {
    private final int mainId;
    private final int extendedId;

    /**
     * The CANMessageId Constructor.
     */
    public CANMessageId(int mainId, int extendedId) {
        this.mainId = mainId;
        this.extendedId = extendedId;
    }

    /**
     * Gets the main id. Bits: {28, 27, 26, ... 0}.
     */
    @Override
    public int getMainId() {
        return this.mainId;
    }

    /**
     * Gets the extended id. Bits: {31, 30, 29}.
     */
    @Override
    public int getExtendedId() {
        return this.extendedId;
    }

    /**
     * Gets the id.
     */
    @Override
    public int getId() {
        return this.mainId | this.extendedId;
    }

    @Override
    public String toString() {
        int id = this.getId();
        return String.valueOf(id);
    }
}

File: DevEnvironmentCLion/CANMessageGenerator/core/CANMessageId.cpp
#include "PreCompiled.h"
#include "CANMessageId.h"

using namespace controllerareanetweok::messagegenerator;

/**
 * Creates a new messages id.
 */
ICANMessageIdPtr CANMessageId::make(unsigned int mainId, unsigned int extendedId)
{
    return std::make_shared<CANMessageId>(mainId, extendedId);
}

/**
 * The CANMessageId Constructor.
 */
CANMessageId::CANMessageId(unsigned int mainId, unsigned int extendedId) :
    m_mainId(mainId),
    m_extendedId(extendedId)
{
}

/**
 * The CANMessageId destructor.
 */
CANMessageId::~CANMessageId()
{
}

/**
 * Gets the main id. Bits: {28, 27, 26, ... 0}.
 */
unsigned int CANMessageId::getMainId() const
{
    return m_mainId;
}

/**
 * Gets the extended id. Bits: {31, 30, 29}.
 */
unsigned int CANMessageId::getExtendedId() const
{
    return m_extendedId;
}

/**
 * Gets the id.
 */
unsigned int CANMessageId::getId() const
{
    return m_mainId | m_extendedId;
}

/**
 * Gets the string representation of this instance.
 */
std::string CANMessageId::toString() const
{
    unsigned int id = getId();
    return std::to_string(id);
}
